components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java

```diff
@@ -329,4 +403,322 @@ private List<String> getBlockedUserStoreDomainsList() {
+
+        Map<String, String> runtimeParams = getRuntimeParams(context);
+        if (MapUtils.isNotEmpty(runtimeParams)) {
+            String skipPreProcessUsername = runtimeParams
+                    .get(MagicLinkAuthenticatorConstants.SKIP_IDENTIFIER_PRE_PROCESS);
+            if (Boolean.parseBoolean(skipPreProcessUsername)) {
+                persistUsername(context, identifierFromRequest);
+
+                // Since the pre-processing is skipped, user id is not populated.
+                setSubjectInContextWithoutUserId(context, identifierFromRequest);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Optional<String> resolveUserFromMultiAttributeLogin(AuthenticationContext context, String username,
+                                                               Map<String, Object> authProperties)
+            throws InvalidCredentialsException {
+
+        if (MagicLinkServiceDataHolder.getInstance().getMultiAttributeLoginService()
+                .isEnabled(context.getTenantDomain())) {
+            ResolvedUserResult resolvedUserResult = MagicLinkServiceDataHolder.getInstance()
+                    .getMultiAttributeLoginService()
+                    .resolveUser(MultitenantUtils.getTenantAwareUsername(username), context.getTenantDomain());
+            if (resolvedUserResult != null && ResolvedUserResult.UserResolvedStatus.SUCCESS.
+                    equals(resolvedUserResult.getResolvedStatus())) {
+                String tenantAwareUsername = resolvedUserResult.getUser().getUsername();
+                String tenantDomain = resolvedUserResult.getUser().getTenantDomain();
+                username = UserCoreUtil.addTenantDomainToEntry(resolvedUserResult.getUser().getUsername(),
+                        context.getTenantDomain());
+                String userId = resolvedUserResult.getUser().getUserID();
+                persistUser(username, authProperties, context, userId, tenantAwareUsername, tenantDomain);
+                return Optional.ofNullable(userId);
+            } else {
+                throw new InvalidCredentialsException(
+                        MagicLinkAuthErrorConstants.ErrorMessages.USER_DOES_NOT_EXISTS.getCode(),
+                        MagicLinkAuthErrorConstants.ErrorMessages.USER_DOES_NOT_EXISTS.getMessage(),
+                        org.wso2.carbon.identity.application.common.model.User.getUserFromUserName(username));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> resolveUserFromOrganizationHierarchy(AuthenticationContext context,
+                                                                  String tenantAwareUsername, String username,
+                                                                  Map<String, Object> authProperties)
+            throws AuthenticationFailedException {
+
+        if (!preconditionsForResolvingUserFromOrganizationHierarchy(context)) {
```

Comment:
how about a method name that suggests it returns a boolean?

suggestion: canResolveUserFromOrganizationHierarchy